apps/labs/posts/numpy-thread-safe-test-suite.md

```diff
@@ -0,0 +1,333 @@
+
+Finally, running tests in parallel can sometimes result in issues with `warnings`. Any sort of global `warnings` filter or capture can lead to thread-unsafe failures. So, instead it's important to keep `warnings` filters under context managers of some sort.
+
+---
+
+## Part Five: Wrapping Things Up
+
+With these test suite modifications under my belt, there were 2 more things to work on:
+
+1. Figuring out what to do with all the tests that I couldn't fix
+2. Getting this all running under NumPy's CI
+
+### Thread Unsafe Markers
+
+#### Problem
+
+When fixing thread-unsafe tests, there were some tests that were thread-unsafe in some way that made fixing them either very difficult or impossible. Maybe they were specifically testing things that were thread-unsafe, like global modules or docstrings. Maybe they modify environment variables. Maybe they require a lot of memory to run, and will completely and utterly crash the terminal when ran in multiple threads (`wsl --shutdown` started to become my best friend over the course of this internship). Or maybe they were using thread-unsafe functionally that I didn't have the scope to fixing during the course of my internship.
+
+Regardless, we still wanted to make sure these wouldn't cause issues when the test suite is ran under `pytest-run-parallel`.
+
+#### Solution
+
+Thankfully, we already have the solution, the `thread_unsafe` marker! One of my latest PRs involved marking up all the remaining tests that were thread-unsafe. That PR got nearly 100 comments over the course of 2-3 weeks, and led to me making more PRs when I realized some of these tests actually could be fixed.
+
+One thing that required some interesting solutions was marking up the entire `np.f2py` module as thread-unsafe. Of course you could go in and put a marker on every test, but that's a bit tedious. Instead, I tried putting a `conftest.py` file in the f2py testing folder that would mark the tests as thread-unsafe with `pytest_itemcollected`.
+
+```python
+# numpy/f2py/tests/conftest.py
+@pytest.hookimpl(tryfirst=True)
+def pytest_itemcollected(item):
+  item.add_marker(pytest.mark.thread_unsafe(reason="f2py tests are thread-unsafe"))
+```
+
+This worked, sort of. With `spin` it worked fine, but when installing NumPy locally using an editable install (`pip install -e . --no-build-isolation`), for some reason this new conftest file overrides the base conftest file in `numpy/` that the test suite needs. My mentor came up with the clever solution of just using the base conftest file and checking that the test was in the f2py directory. Fun stuff!
+
+I definitely submitted this PR at a bad time, overlapping with the Quansight retreat, so shout out to Nathan, Lysandros, and [Ralf Gommers](https://github.com/rgommers) for taking a look at it despite that!
```

Comment:
I'd just delete this. I don't think anyone did anything wrong, sometimes people have commitments and things take time. It's not anyone's fault.